### ISSUE TITLE
PassivePort randomization is broken due to SO_REUSEPORT option

### ISSUE THREAD

**Issue by Anthrazz** [proftpd/proftpd#1171]
Proftpd does use the same server port for multiple passive FTP connections. Even when executing multiple simultaneous FTP sessions from different clients. This does break simultaneous passive FTP connections, file listings and transfers.

##### FTP Session 1:
```bash
# ftp 111.111.111.111
Connected to 111.111.111.111 (111.111.111.111).
220 ProFTPD Server (ProFTPD) [::ffff:111.111.111.111]
Name (111.111.111.111:xxxxxxxx): username
331 Password required for username
Password:
230 User username logged in
Remote system type is UNIX.
Using binary mode to transfer files.
ftp> ls -ahl
227 Entering Passive Mode (111,111,111,111,255,4).      # <-- the same passive port like in the other session
150 Opening ASCII mode data connection for file list
drwxr-xr-x   2 username  username   3 Jan 21 09:36 .
drwxr-xr-x   2 username  username   3 Jan 21 09:36 ..
226 Transfer complete
ftp> ls -ahl
227 Entering Passive Mode (111,111,111,111,220,128).    # <-- the same passive port like in the other session
150 Opening ASCII mode data connection for file list
drwxr-xr-x   2 username  username   3 Jan 21 09:36 .
drwxr-xr-x   2 username  username   3 Jan 21 09:36 ..
226 Transfer complete
```
FTP Session 2 (simultaneous session but commands executed with some seconds between):
```bash
# ftp 111.111.111.111
Connected to 111.111.111.111 (111.111.111.111).
220 ProFTPD Server (ProFTPD) [::ffff:111.111.111.111]
Name (111.111.111.111:xxxxxxxx): username
331 Password required for username
Password:
230 User username logged in
Remote system type is UNIX.
Using binary mode to transfer files.
ftp> ls -ahl
227 Entering Passive Mode (111,111,111,111,255,4).    # <-- the same passive port like in the other session
150 Opening ASCII mode data connection for file list
drwxr-xr-x   2 username  username   3 Jan 21 09:36 .
drwxr-xr-x   2 username  username   3 Jan 21 09:36 ..
226 Transfer complete
ftp> ls -ahl
227 Entering Passive Mode (111,111,111,111,220,128).    # <-- the same passive port like in the other session
150 Opening ASCII mode data connection for file list
drwxr-xr-x   2 username  username   3 Jan 21 09:36 .
drwxr-xr-x   2 username  username   3 Jan 21 09:36 ..
226 Transfer complete
```

* if a second command is executed another port will be used but it is not choosen randomly
* if commands are executed simultaneous, one usually succeeds while the remaining sessions breaks

### How to Workaround

Without the configuration option `PassivePorts` the randomization is working correctly because the Kernel does assign the port.

### What I Expected/Wanted

Using an randomly choosen port for each FTP command for the passive FTP connection.

### ProFTPD Version and Configuration

```bash
# proftpd -V
Compile-time Settings:
  Version: 1.3.7a (maint)
  Platform: FREEBSD11 (FREEBSD11_4) [FreeBSD 11.4-RELEASE-p5 amd64]
  Built: Thu Jan 14 2021 12:48:37 UTC
  Built With:
    configure  '--localstatedir=/var/run' '--libexecdir=/usr/local/libexec/proftpd' '--with-pkgconfig=libdata/pkgconfig' '--sysconfdir=/usr/local/etc' '--enable-ctrls' '--enable-dso' '--disable-sendfile' '--enable-ipv6' '--disable-memcache' '--enable-nls' '--enable-pcre' '--disable-redis' '--with-shared=mod_ban:mod_copy:mod_ctrls_admin:mod_deflate:mod_dnsbl:mod_dynmasq:mod_exec:mod_ifsession:mod_ifversion:mod_qos:mod_quotatab:mod_quotatab_file:mod_quotatab_radius:mod_quotatab_sql:mod_radius:mod_ratio:mod_readme:mod_rewrite:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_site_misc:mod_snmp:mod_sql:mod_sql_passwd:mod_tls:mod_tls_shmcache:mod_unique_id:mod_wrap2:mod_wrap2_file:mod_wrap2_sql' '--with-includes=/usr/local/include' '--with-libraries=/usr/local/lib' '--prefix=/usr/local' '--mandir=/usr/local/man' '--infodir=/usr/local/share/info/' '--build=amd64-portbld-freebsd11.4' 'build_alias=amd64-portbld-freebsd11.4' 'CC=cc' 'CFLAGS=-O2 -pipe  -DLIBICONV_PLUG -fstack-protector-strong -fno-strict-aliasing ' 'LDFLAGS= -lpthread -Wl,-rpath,/usr/local/lib -fstack-protector-strong ' 'LIBS=-lssl -lcrypto -L/usr/local/lib' 'CPPFLAGS=-DHAVE_OPENSSL -I/usr/local/include -DLIBICONV_PLUG' 'CPP=cpp' 'CXX=c++' 'CXXFLAGS=-O2 -pipe -DLIBICONV_PLUG -fstack-protector-strong -fno-strict-aliasing  -DLIBICONV_PLUG '

  CFLAGS: -g2 -O2 -pipe  -DLIBICONV_PLUG -fstack-protector-strong -fno-strict-aliasing  -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration
  LDFLAGS: -L$(top_srcdir)/lib -L$(top_builddir)/lib  -lpthread -Wl,-rpath,/usr/local/lib -fstack-protector-strong  -rdynamic -L/usr/local/lib
  LIBS: -lintl  -lpcreposix -lpcre -lssl -lcrypto -lpam -lsupp -lexecinfo -lcrypt -lssl -lcrypto -L/usr/local/lib -lutil -pthread

  Files:
    Configuration File:
      /usr/local/etc/proftpd.conf
    Pid File:
      /var/run/proftpd.pid
    Scoreboard File:
      /var/run/proftpd.scoreboard
    Header Directory:
      /usr/local/include/proftpd
    Shared Module Directory:
      /usr/local/libexec/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    - Autoshadow support
    + Controls support
    + curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    - Lastlog support
    - Memcache support
    + ncursesw support
    + NLS support
    + OpenSSL support (OpenSSL 1.1.1i  8 Dec 2020)
    + PCRE support
    - POSIX ACL support
    - Redis support
    - Sendfile support
    - Shadow file support
    - Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 1024
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

```bash
# cat /usr/local/etc/proftpd.conf
ServerType                        standalone

User                              nobody
Group                             nogroup

Port                              21

RequireValidShell                 off
DefaultRoot                       ~
AllowChrootSymlinks               off

PassivePorts                      49152 65534
```

**Comment by Castaglia** [comment]
Can you provide the ProFTPD debug logging, debug level 10, generated when you encounter these issues, please?  See http://www.proftpd.org/docs/howto/Debugging.html

**Comment by Anthrazz** [comment]
I have attached the logs and also the trace log of the Proftpd. The trace log was generated with the following configuration:
```
TraceLog /backup/log/proftpd-debug-log/trace.log
Trace DEFAULT:20 signal:0
```

In this test both FTP list were executed at the same time. The first session has worked and the second has hanged and must be killed manually on the client.

[proftpd_debug.txt](https://github.com/proftpd/proftpd/files/5865550/proftpd_debug.txt)
[proftpd_trace.log](https://github.com/proftpd/proftpd/files/5865540/proftpd_trace.log)

[ftp_session1.txt](https://github.com/proftpd/proftpd/files/5865543/ftp_session1.txt)
[ftp_session2.txt](https://github.com/proftpd/proftpd/files/5865542/ftp_session2.txt)

**Comment by Castaglia** [comment]
I believe this behavior is occurring because of the use of the `SO_REUSEPORT` socket option [here](https://github.com/proftpd/proftpd/blob/master/src/inet.c#L334).  Another user, independently, reported this same issue to me earlier this week.

The original [commit](https://github.com/proftpd/proftpd/commit/b69cf7483a78ec913db03d28c93517dc0fc6fe58) for that `SO_REUSEPORT` option does not provide enough context (that's my fault).  And without that context, in the face of these issues, I'm just going to remove it.  We can add it back later, properly and with context, if ever needed.

**Comment by Castaglia** [comment]
Hmm.  Turns out that there _is_ a need for `SO_REUSEPORT`; see https://github.com/Castaglia/proftpd-mod_proxy/issues/122#issuecomment-331982920.

It's the _implementation_ that is causing the problems.  That is, using `SO_REUSEPORT` may be acceptable, but _only_ for *active* data transfers.  Not for the control connection, and not for passive data transfers.

**Comment by Castaglia** [comment]
This should be fixed in master now; the fix has been backported to the 1.3.7 branch as well.  Thanks!
